src/foam/nanos/analytics/DAOReduceManager.js

```diff
@@ -42,19 +61,18 @@ foam.mlang.sink.GroupBy dataToReduce = (foam.mlang.sink.GroupBy) getSourceDAO()
   .select(GROUP_BY(Candlestick.KEY, new foam.dao.ArraySink.Builder(x).build()));
 
 for ( Object key : dataToReduce.getGroups().keySet() ) {
-  java.util.Map<Long, Candlestick> reducedData = new java.util.HashMap<Long, Candlestick>();
+  java.util.Map<Date, Candlestick> reducedData = new java.util.HashMap<>();
   for ( Object o : ((foam.dao.ArraySink) dataToReduce.getGroups().get(key)).getArray() ) {
     Candlestick c = (Candlestick) o;
-    long cCloseTimeMs = c.getCloseTime().getTime();
-    long rCloseTimeMs = cCloseTimeMs - ( cCloseTimeMs % periodMs ) + periodMs;
-    if ( ! reducedData.containsKey(rCloseTimeMs) ) {
-      reducedData.put(rCloseTimeMs, new Candlestick.Builder(x)
-        .setOpenTime(new Date(rCloseTimeMs - periodMs))
-        .setCloseTime(new Date(rCloseTimeMs))
+    Date rCloseTime = (Date) getCloseTimeExpr().f(c.getCloseTime());
+    if ( ! reducedData.containsKey(rCloseTime) ) {
```

Comment:
Can we make Candlestick a sink? Or a CandlestickSink so that we don't have to do this looping ourselves?  Would be much better if we just wrote.

```
getSourceDAO().where().select(GROUP_BY(Candlestick.KEY, CANDLESTICK()))
```

Where CANDLESTICK is a sink which creates a new candelstick on first put, and reduces into the existing one on every subsequent put.

Reply:
Yeah I like how you're thinking about this but I'm thinking we do that in a follow-up PR. This PR opens the door to being able to do monthly/quarterly candlesticks and we'll follow this up with better naming of our interfaces/methods for this stuff and a sink or two for folding and reducing.